src/notation/view/internal/dynamicpopupmodel.cpp

```diff
@@ -0,0 +1,225 @@
+    },
+    {   // Page 6
+        { DynamicType::PPPPP,  64, 2.0, DynamicPopupModel::Dynamic,
+          "<sym>dynamicPiano</sym><sym>dynamicPiano</sym><sym>dynamicPiano</sym><sym>dynamicPiano</sym><sym>dynamicPiano</sym>" },
+        { DynamicType::PPPP,   52, 2.0, DynamicPopupModel::Dynamic,
+          "<sym>dynamicPiano</sym><sym>dynamicPiano</sym><sym>dynamicPiano</sym><sym>dynamicPiano</sym>" },
+        { DynamicType::PPP,    44, 2.0, DynamicPopupModel::Dynamic,
+          "<sym>dynamicPiano</sym><sym>dynamicPiano</sym><sym>dynamicPiano</sym>" },
+    },
+    {   // Page 7 - Hairpins
+        { DynamicType::OTHER,  62, 0.0, DynamicPopupModel::Crescendo, "" },
+        { DynamicType::OTHER,  62, 0.0, DynamicPopupModel::Decrescendo, "" },
+    },
+};
+
+DynamicPopupModel::DynamicPopupModel(QObject* parent)
+    : AbstractElementPopupModel(PopupModelType::TYPE_DYNAMIC, parent)
+{
+}
+
+QString DynamicPopupModel::fontFamily() const
+{
+    return QString::fromStdString(m_item->score()->engravingFont()->family());
+}
+
+QVariantList DynamicPopupModel::pages() const
+{
+    return m_pages;
+}
+
+QString DynamicPopupModel::xmlTextToQString(const std::string text, IEngravingFontPtr engravingFont) const
+{
+    const std::string startTag = "<sym>";
+    const std::string endTag = "</sym>";
+
+    String str;
+    size_t pos = 0;
+
+    while (true) {
+        size_t startPos = text.find(startTag, pos);
+        if (startPos == std::string::npos) {
+            break;
+        }
+        size_t symNameStart = startPos + startTag.length();
+        size_t endPos = text.find(endTag, symNameStart);
+        if (endPos == std::string::npos) {
+            break;
+        }
+        SymId id = SymNames::symIdByName(text.substr(symNameStart, endPos - symNameStart));
+        str += engravingFont->toString(id);
+        pos = endPos + endTag.length();
+    }
+
+    return str.toQString();
+}
+
+void DynamicPopupModel::init()
+{
+    AbstractElementPopupModel::init();
+
+    m_pages.clear();
+
+    if (!m_item) {
+        return;
+    }
+
+    IEngravingFontPtr engravingFont = m_item->score()->engravingFont();
+
+    for (const QList<DynamicPopupModel::PageItem>& page : DYN_POPUP_PAGES) {
+        QVariantList variantPage;
+        for (const DynamicPopupModel::PageItem& item : page) {
+            QVariantMap variantMap {
+                { "text", xmlTextToQString(item.text, engravingFont) },
+                { "width", item.width },
+                { "offset", item.offset },
+                { "type", item.itemType },
+            };
+            variantPage.append(variantMap);
+        }
+        m_pages.append(QVariant::fromValue(variantPage));
+    }
+
+    emit pagesChanged();
+}
+
+void DynamicPopupModel::addOrChangeDynamic(int page, int index)
+{
+    IF_ASSERT_FAILED(m_item) {
+        return;
+    }
+
+    if (!m_item->isDynamic()) {
```

Comment:
Is this a condition that can validly occur? Or should you maybe consolidate it with the assertion above? 
```c++
    IF_ASSERT_FAILED(m_item && m_item->isDynamic()) {
        return;
    }
```

Reply:
I added it as safety check, not sure it will occur. We can consolidate in the assertion👍